python/cudf/cudf/core/dataframe.py

```diff
@@ -650,6 +650,11 @@ def __init__(
                     data, index=index, columns=columns, nan_as_null=nan_as_null
                 )
 
+        if self._data.nrows > 0 and self._data.nrows != len(self._index):
```

Comment:
Are we introducing this check for extra validation

Reply:
This check is necessary. Previously we were assigning the indexes in the constructor via the property `self.index = ...`. Now we are using `self._index = ...`, which won't do the same error checking. The reason for the change is that the index property setter uses `len(self)` to check the mismatch, which in turn calls `self._num_rows()`. Previously all Frames would [check whether the index is None when requesting this](https://github.com/rapidsai/cudf/pull/10439/files#diff-45f163ddc7fb3fc5853e1fdd90a3d740a9dd0987c9ebcf00315ad7aff727576dL147). In practice, that is not necessary for IndexedFrame objects because the index is never None, _except_ during initialization. At this stage the object is not yet in a valid state, so we cannot safely use the property and we need to do the check manually in the constructor. Eventually I'd like to refactor the constructor and simplify this, but for now I think the duplication is the best that we can do.

Reply:
Discussed offline, we want to introduce a helper function that is only called in the cases where it's needed to minimize unnecessary validation.